velox/exec/Spill.h

```diff
@@ -0,0 +1,340 @@
+  // the target number of rows in a single RowVector written to a spill file.
+  // 'pool' and 'mappedMemory' own the memory for state and results.
+  SpillState(
+      TypePtr type,
+      const std::string& path,
+      HashBitField bits,
+      uint64_t targetFileSize,
+      uint64_t targetBatchSize,
+      memory::MemoryPool& pool,
+      memory::MappedMemory* mappedMemory)
+      : type_(type),
+        path_(path),
+        bits_(bits),
+        fieldMask_(((1UL << (bits_.end - bits_.begin))) - 1),
+        targetFileSize_(targetFileSize),
+        targetBatchSize_(targetBatchSize),
+        pool_(pool),
+        mappedMemory_(mappedMemory) {}
+
+  // Returns which spill partition 'hash' falls. Returns -1 if the
+  // partition of 'hash' is has not been started.
+  int32_t way(uint64_t hash) {
+    auto field = (hash >> bits_.begin) & fieldMask_;
+    return field < numWays_ ? field : -1;
+  }
+
+  int32_t numWays() const {
+    return numWays_;
+  }
+
+  // Sets how many of the spill partitions are in use.
+  void setNumWays(int32_t numWays);
+
+  // Returns how many ways spilled data can be partitioned.
+  uint16_t maxWays() {
+    return 1 << (bits_.end - bits_.begin);
+  }
+
+  uint64_t targetFileSize() const {
+    return targetFileSize_;
+  }
+
+  uint64_t targetBatchSize() const {
+    return targetBatchSize_;
+  }
+
+  memory::MemoryPool& pool() {
+    return pool_;
+  }
+
+  // Appends data to 'way'. The rows  given by 'indices'  must be sorted and
+  // must hash to 'way'.
+  void write(
+      uint16_t way,
+      RowVectorPtr rows,
+      const folly::Range<IndexRange*> indices);
+
+  // Finishes a sorted run for 'way'. If write is called for 'way'
+  // again, the data does not have to be sorted relative to the data
+  // written so far.
+  void finishWrite(uint16_t way) {
+    files_[way]->flush(true);
+  }
+
+  // Starts reading values for 'way'. If 'extra' is non-null, it can be
+  // a stream of rows from a RowContainer so as to merge unspilled
+  // data with spilled data.
+  std::unique_ptr<TreeOfLosers<VectorRow, SpillStream>> startMerge(
+      uint16_t way,
+      std::unique_ptr<SpillStream>&& extra);
+
+  // Helper function for comparing current elements of streams to merge.
+  static int32_t compareSpilled(
+      const VectorRow& left,
+      const VectorRow& right,
+      int32_t numKeys);
+
+ private:
+  const TypePtr type_;
+  const std::string path_;
+  const HashBitField bits_;
+  const uint64_t fieldMask_;
+  // Number of spilled ranges. All hashes where the 'bits_' <
+  // numSpilledRanges spill to the file list in the corresponding
+  // place in 'files_'.
+  int32_t numWays_ = 0;
+  const uint64_t targetFileSize_;
+  const uint64_t targetBatchSize_;
+  // A file list for each spilled range.
+  std::vector<std::unique_ptr<FileList>> files_;
+  memory::MemoryPool& pool_;
+  memory::MappedMemory* const mappedMemory_;
+  static uint64_t sequence_;
+  static std::mutex mutex_;
```

Comment:
Would you document what is this mutex for? Which data is it protected? Can we use folly::Synchronized instead? For some reason I cannot find where sequence_ or mutex_ are used. I'd appreciate a pointer.